.travis.yml

```diff
@@ -13,74 +18,118 @@ python:
+
+  # Install geckodriver required for selenium testing
+  - |
+    if [ "$REFINERY_SELENIUM_TESTS" ] || [ "$TOOL_MANAGER_SELENIUM_TESTS" ]; then
+      GECKODRIVER_VERSION=v0.17.0
+      TAR=geckodriver-$GECKODRIVER_VERSION-linux64.tar
+      wget https://github.com/mozilla/geckodriver/releases/download/$GECKODRIVER_VERSION/$TAR.gz
+      gunzip $TAR.gz
+      tar -xvf $TAR
+      chmod a+x geckodriver
+      export PATH=$PATH:`pwd`
+    fi
 
 before_script:
-  # Vagrant is not used on travis: This user is just for the database.
-  - createuser --createdb --no-superuser --no-createrole vagrant
-  - createdb -O vagrant refinery
-
-  - sudo mkdir /data # In vagrant, this is created by puppet; In production, this is an EBS mount.
-  - sudo chown $USER /data
-
-  - cd refinery
-# See http://www.stuartellis.eu/articles/erb/#running-erb-from-the-command-line
-  - erb config/config.json.erb > config/config.json
-  - python manage.py migrate --noinput
-  - npm install -g grunt-cli@0.1.13 bower@1.7.7 --progress false --quiet || ( cat npm-debug.log && false )
-  - cd ui
-  - npm install --progress false --quiet || ( cat npm-debug.log && false )
-  - bower install --config.interactive=false --quiet
-  - cd ../
+  - |
+    if ! [ "$GRUNT_TESTS" ]; then
+      # Vagrant is not used on travis: This user is just for the database.
+      createuser --createdb --no-superuser --no-createrole vagrant
+      createdb -O vagrant refinery
+      sudo mkdir /data
+      sudo chown $USER /data
+      pushd refinery
+      # See http://www.stuartellis.eu/articles/erb/#running-erb-from-the-command-line
+      erb config/config.json.erb > config/config.json
+      python manage.py migrate --noinput
+      popd
+    fi
+
+  - |
+    if ! [ "$DJANGO_UNIT_TESTS" ]; then
+      pushd refinery
+      npm install -g grunt-cli@0.1.13 bower@1.7.7 --progress false --quiet || ( cat npm-debug.log && false )
+      pushd ui
+      npm install --progress false --quiet || ( cat npm-debug.log && false )
+      bower install --config.interactive=false --quiet
+      popd && popd
+    fi
+
 script:
   - set -e # Any error will cause travis to exit early and report a failure.
+  - start $FLAKE8
+  - |
+    if ! [ "$GRUNT_TESTS" ]; then
+      # TODO: Not the same behavior as pre-commit hook.
+      # https://github.com/refinery-platform/refinery-platform/issues/1851
+      flake8 --exclude=migrations,ui ..
+    fi
+  - end $FLAKE8
 
-  # TODO: Not the same behavior as pre-commit hook.
-  # https://github.com/refinery-platform/refinery-platform/issues/1851
-  - flake8 --exclude=migrations,ui ..
+  - start $GRUNT
+  - pushd refinery
+  - pushd ui
+  - if ! [ "$DJANGO_UNIT_TESTS" ]; then grunt make; fi
+  - if [ "$GRUNT_TESTS" ]; then grunt test; fi
+  - popd
+  - end $GRUNT
 
-  - echo 'travis_fold:start:grunt'
-  - pushd ui && grunt && popd
-  - echo 'travis_fold:end:grunt'
+  - start $DJANGO_UNIT_TESTS
+  - if [ "$DJANGO_UNIT_TESTS" ]; then coverage run manage.py test; fi
```

Comment:
`manage.py test` runs all tests, including Selenium: https://travis-ci.org/refinery-platform/refinery-platform/jobs/286079823

It would be great to move **all** Selenium tests into a separate folder (`selenium_tests`?) and [name them such that they wouldn't be picked up](https://docs.djangoproject.com/en/1.7/topics/testing/overview/#running-tests) by `manage.py test` command. Then they can be easily run separately by something like `manage.py test -p="selenium_*.py`.

Reply:
@hackdna `manage.py test` by default only runs python files matching the pattern `test*.py`, so the selenium tests are not executed by that command now. Only our unit tests are.

In this PR Selenium-based tests have been moved to files named `selenium_tests.py` in their respective Django apps, and we are running them in separate builds ussing the`--pattern` already.

The `selenium_testing` app has tests specific to Refinery UI code, but I felt like it would be odd to include the `tool_manager` ones there as well since we are utilizing Selenium in those to ensure that Refinery plays nicely with containerized visualizations, not testing Refinery UI code

Reply:
I see. This should be documented because it is not obvious why `selenium_testing/tests.py` are not discovered by default.

The purpose of unit testing is to test the smallest testable parts of an application, so they should not have external dependencies and should run as quickly as possible to enable quick development cycle. On the other hand, integration/functional tests are not limited to just testing of the UI code. They test how application components (and their dependencies) work together.

Existing tool manager tests have external dependencies and test several components at once. So, they are actually integration/functional tests and thus should be grouped together with other such tests.

Reply:
`selenium_testing/tests.py` doesn't exist now. It is now `selenium_testing/selenium_tests.py`

And sure that sounds good, but we have integration tests mixed in with our unit tests throughout our codebase so I'm not sure if that type of change is in the scope of this PR.

Reply:
Still, it is not immediately obvious that the name of that file is significant. I would just put a note at the top of it and mention this on the wiki page about testing.

There are not too many of those tests and moving them is straightforward. I think this PR provides a great context for making this change.